### ISSUE TITLE
Auto-indent doesn't respect VS Code indentation settings in new files

### ISSUE THREAD

**Issue by GinoMan** [microsoft/pylance-release#4071]
Type: <b>Bug</b>

# Behaviour
## Expected vs. Actual

Expected Behavior:
When hitting return to go to the next line that is auto-indented, the indentation should follow your own tabs vs spaces configuration.

Actual Behavior:
Typing in any line whose next line should be indented and hitting enter inserts 4 spaces instead of following your user and workspace configuration

## Steps to reproduce:

1. Install Python Extension
2. Restart Visual Studio Code
3. Disable linting, put in either or both the project and workspace configuration the following options inside or outside a "[python]" key:

```json
"python.pipenvPath": "./.venv/Scripts/pipenv.exe",
"editor.tabSize": 4,
"editor.insertSpaces": false,
"[python]": {
	"editor.tabSize": 4,
	"editor.insertSpaces": false,
	"editor.indentSize": "tabSize",
	"editor.guides.indentation": true,
	"editor.detectIndentation": false,
	"files.trimTrailingWhitespace": true,
},
"python.linting.enabled": false,
```

4. Create a new python file
5. Type in any statement that expects the next line to be indented. Examples include...

```python
if __name__ == "__main__":
# or
def main_function():
# or
while True:
# or
class MyClass:
```

7. Hit the enter key
8. Open the command palette and type in "render whitespace" and toggle the editor to show them.
9. Observe that instead of a tab as configured, you have four spaces at the beginning of the next line.

![VSCode Python Extension Bug](https://user-images.githubusercontent.com/528834/224846682-ba01c813-bc94-42a6-a9b0-cd16e2140b17.gif)

# Diagnostic data

-   Python version (& distribution if applicable, e.g. Anaconda): 3.11.0
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): PipEnv
-   Value of the `python.languageServer` setting: Default

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
(Note, nothing is printed to the output console when this happens either.
```

</p>
</details>

<details>

<summary>User Settings</summary>

<p>

```

pipenvPath: "<placeholder>"

languageServer: "Pylance"

linting
• enabled: false

```

</p>
</details>

Extension version: 2023.4.1
VS Code version: Code 1.76.1 (5e805b79fcb6ba4c2d23712967df89a089da575b, 2023-03-08T16:32:00.131Z)
OS version: Windows_NT x64 10.0.19044
Modes:
Sandboxed: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz (4 x 2904)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|undefined|
|Memory (System)|23.89GB (10.86GB free)|
|Process Argv|--folder-uri file:///c%3A/Users/OpenM/source/repos/Edit-Env --crash-reporter-id cb928791-934b-4d70-ac94-10d42d342b49|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vsclangdf:30486550
c4g48928:30535728
dsvsc012cf:30540253
pynewvextcfv2:30681851
azure-dev_surveyone:30548225
pyindex848:30662994
nodejswelcome1cf:30587006
3biah626:30602489
pyind779:30671433
f6dab269:30613381
vscrp:30673768
pythonsymbol12:30671437
a9j8j154:30646983
showlangstatbar:30672706
vsctsb:30677850
funwalk2cf:30682975
pythonms35cf:30680985

```

</details>

**Comment by GinoMan** [comment]
Weird. I couldn't find this on my computer.

**Comment by DelofJ** [comment]
Just to say, I also have this issue, pretty annoying

**Comment by debonte** [comment]
Our auto-indent system is currently not honoring non-default VS Code indentation settings. This should primarily be an issue for new files. In existing files with at least a few lines and indents, we will detect the most prevalent indentation type and use that.

You can get the old behavior by setting:

```json
    "[python]": {
        "editor.formatOnType": false
    }
```

**Comment by SunsetOrange** [comment]
I just tracked down the source of this rage inducing behaviour to Pylance myself, which led me to this issue. I was going to raise an issue, but I'm glad to see I am not the only one with the problem.
However, I will point out that this isn't just new files. I get the same behaviour with existing files.
> "editor.formatOnType": false

Does fix the issue, but I'm wondering what other features does it disable that I'd rather keep.

**Comment by debonte** [comment]
> I get the same behaviour with existing files.

@SunsetOrange, can you provide a file that repros this behavior along with your settings? With an existing file (more than one line), we should use whatever line ending is most prevalent in the file.

> I'm wondering what other features does it disable that I'd rather keep.

[This comment](https://github.com/microsoft/pylance-release/issues/4378#issuecomment-1551786406) describes the two features that would currently be impacted by disabled `editor.formatOnType`.

**Comment by GinoMan** [comment]
> > I'm wondering what other features does it disable that I'd rather keep.
> 
> [This comment](https://github.com/microsoft/pylance-release/issues/4378#issuecomment-1551786406) describes the two features that would currently be impacted by disabled `editor.formatOnType`.

Yeah, pretty sure I'd like to keep those. I don't want to have to hit tab every time or backspace every time I want to change indentations. Also the fstring prefix is a nice feature.

**Comment by SunsetOrange** [comment]
> @SunsetOrange, can you provide a file that repros this behavior along with your settings? With an existing file (more than one line), we should use whatever line ending is most prevalent in the file.

So I found the reason. It only detects indentation type if the indentation is used to define indented scoped code, like in a function or a conditional statement. The indentation type is not detected if your code only has indentation within variable definitions.

For instance, indentation as only code formatting within variable definitions.
![Screenshot 2023-05-28 121044](https://github.com/microsoft/pylance-release/assets/7814343/ca020235-86c3-4daf-8ebe-2b54d3dba31c)

However, indentation defining code scopes.
![Screenshot 2023-05-28 121124](https://github.com/microsoft/pylance-release/assets/7814343/db3a7139-73a0-406e-bb5d-3b33ab9631c4)
(cursor not visable)

settings.json was...
```
{
	"editor.tabSize": 4,
	"editor.insertSpaces": false,
	"editor.detectIndentation": false,
	// "[python]": {
	// 	"editor.formatOnType": false
	// },
}
```

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.8.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023841-23-August-2023-PreRelease)
